### ISSUE TITLE
Better Extensibility Model For Frameworks

### ISSUE THREAD

**Issue by ryanjbaxter** [dekorateio/dekorate#286]
This might be applicable to other frameworks, but my only experience is with Spring.

What we do now is add some logic in `SpringBootApplicationProcessor` to add certain properties to the manifest we generate based on whether certain starters are on the classpath (for example Spring Cloud Kubernetes, Spring Actuator).  There are many many Spring starters in the wild and it would be nice if we didn't have to add everything to this one class and also if the code could live outside of Dekorate itself.  The ideal situation in my opinion would be if we could have the Dekorate Spring starter on the classpath and then any number of other starters that allow us to further "decorate" the manifest based on their needs.  @iocanel @geoand thoughts?

**Comment by iocanel** [comment]
@ryanjbaxter I totally agree. 

That was the main drive for the BeanListener class which was recently introduce.
We can take it even further and add a class listener that can check if certain classes are present on the classpath.

At least for now I don't see some other recurring pattern. What do you think?

**Comment by ryanjbaxter** [comment]
@iocanel I think that is a nice start but not really what I am envisioning.  I would like these "listeners" to live outside of Dekorate in other jars.  Dekorate would search for instances of these listener interfaces and then use the ones it finds.  Does that make sense?

**Comment by ryanjbaxter** [comment]
Quoting @iocanel from #288

> Currently, using the java.util.ServiceLoader anyone can provide a custom BeanListener which can dekorate the manifests as sees fit. We can extend this concept to classes too (as mentioned in #286).
> 
> We may also need to find a way to override a bean or a class listener to change the behavior of a known listener (not 100% sure about it, needs some more thought).
> 
> The big question is what's the ideal place for these listeners. Are we to bundle them inside the dekorate spring module, or we should push them to each project individuallly (e.g. move the sck integration in sck itself and so on), or even somewhere else?
> 

Originally I was thinking of having a Spring Dekorate project that we (the Spring team) own.  

> Personally, as long as there is concensus on what is best for the majority of users, I'd rather keep most of listeners in the existing spring module. The main reason is that dekorate is a small project and at this stage it has quite a fast release cycle so it can help us move faster. An other reason is that it easier to maintain (less modules, less things to depend on dekorate), which also means easier for the user to get started (nothing can beat the drop a single jar to your classspath experience).
 
I agree that Dekorate has been fine as far as contributing code and evolving quickly.  My only thought it that Spring has a such a big feature set (not that everything is going to have integration with Dekorate) that it might be a bit much.

> So what I propose is the following, let's design together clear extension points. Let's provide a set of sensible defaults as part of the dekorate spring module. And from there if need be anyone can create specialized modules.
> 
> wdyt?

Maybe this makes sense to start with.  If we need to we can then separate that out into its own project if it evolves in a way that makes sense.

@spencergibb wondering if you have any opinion one way or the other here

**Comment by spencergibb** [comment]
I don't think I like the idea of us managing a dedicated Spring Dekorate project. I think I like the idea of using something like `SpringFactoryLoader` which is how all spring boot auto configuration and cloud bootstrap work.

https://github.com/spring-cloud/spring-cloud-commons/blob/352bc60cdf9ecedb1d9bc8ef126e20c3aa0e39ec/spring-cloud-context/src/main/java/org/springframework/cloud/bootstrap/BootstrapApplicationListener.java#L163-L164

That basic framework cloud live in https://github.com/dekorateio/dekorate/tree/master/frameworks/spring-boot. Any project that would want to plug in could just add an entry in `META-INF/spring.factories`.

Maybe the service loader is enough, but any project should be able to add to spring dekorate.

**Comment by iocanel** [comment]
I don't think we can use spring factory loader as everything we do happens during compile time. That's why I went with j.u.ServiceLoader.

**Comment by spencergibb** [comment]
Yeah, I realized that as I was writing it. @ryanjbaxter what issue do you have with service loader?

**Comment by ryanjbaxter** [comment]
I dont have any issue :)  I just wanted some other eyes on the subject that is all.  I think the `ServiceLoader` is a fine place to start.
